design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,263 @@
+| `--enable-certificate-owner-ref` not passed to the controller | No change needed |
+| `--enable-certificate-owner-ref=false` | Replace with `--default-secret-cleanup-policy=Never` |
+| `--enable-certificate-owner-ref=true` | Replace with `--default-secret-cleanup-policy=OnDelete` |
+
+## Design Details
+
+cert-manager would have to change in a few places.
+
+**Mutating webhook**
+
+We propose to have no "value defaulting" for `cleanupPolicy` because the
+"empty" value has a meaning for us: when `cleanupPolicy` is empty, the
+presence or not of the flag `--enable-certificate-owner-ref` takes over.
+To give more context, some other resources, such as the Pod resource,
+will mutate the object when the value is "empty", for example the
+`imagePullPolicy` value will default to `IfNotPresent`.
+
+**PostIssuancePolicyChain**
+
+In ([policies.go#L95](https://github.com/cert-manager/cert-manager/blob/b78af1ef867f8776715cae3dd6a8b83049c4d9b2/internal/controller/certificates/policies/policies.go#L95-L104)), cert-manager does a few sanity checks right after the issuer (either an
+internal or an external issue) has filled the CertificateRequest's status
+with the signed certificate.
+
+One of the checks is called
+[`SecretOwnerReferenceValueMismatch`](https://github.com/cert-manager/cert-manager/blob/b78af1ef867f8776715cae3dd6a8b83049c4d9b2/internal/controller/certificates/policies/checks.go#L511)
+and checks that the owner reference on the Secret resource matches the one
+on the Certificate resource.
+
+### Test Plan
+
+- Unit tests for the changes in the secret manager controller.
+- Integration tests (either fake client or envtest) checking various API behaviours.
+
+### Graduation Criteria
+
+We propose to release this feature in GA immediately and skip the "beta"
+phase that consists of gathering user feedback, since this feature has a
+low user-facing surface. We think that we will be able to take a good
+decision (e.g., the name of the new field, whether it is a boolean or a
+string, and which values the field can take) while developing the feature
+in the PR.
+
+We don't think this feature needs to be [feature gated][feature gate].
+
+[feature gate]: https://cert-manager.io/docs/installation/featureflags/#list-of-current-feature-gates
+
+### Upgrade / Downgrade Strategy
+
+Upgrading from a version without this feature to a version with this
+feature won't be breaking.
+
+Downgrading, however, will be a breaking change, since a new field will be
+introduced.
+
+### Supported Versions
+
+This feature will be supported in all the versions of Kubernetes that are supported by cert-manager.
+
+## Alternatives
+
+**CSI driver**
+
+It is possible to use the
+[`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
+the problem of "too many ephemeral Secret resources stored in etcd". Using
+the CSI driver, no Secret resource is created, alleviating the issue. Since Flant offers its customers the capability to use Certificate resources,
```

Comment:
```suggestion
It is possible to use a
[`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
the problem of "too many ephemeral Secret resources stored in etcd". Using
a CSI driver, no Secret resource is created, alleviating the issue. Since Flant offers its customers the capability to use Certificate resources,
```